website/content/docs/concepts/username-templating.mdx

```diff
@@ -0,0 +1,143 @@
+---
+layout: docs
+page_title: Username Templating
+sidebar_title: Username Templating
+description: >-
+  Username templating are used in some secret engines to allow operators to define
+  how dynamic usernames are generated.
+---
+
+# Username Templating
+
+Some of the secrets engines that generate dynamic users for external systems provide the ability for Vault operators
+to customize how usernames are generated for said external systems. This uses the
+[Go template language](https://golang.org/pkg/text/template/). This page describes the basics of using these templates
+for username generation but does not go into great depth of using the templating language for more advanced usages.
+See the API documentation for the given secret engine to determine if it supports username templating and for more
+details on using it with that engine.
+
+~> When customizing how usernames are generated, take care to ensure you have enough randomness to ensure uniqueness
+   otherwise multiple calls to create the credentials may interfere with each other.
+
+In addition to the functionality built into the Go template language, a number of additional functions are available:
+
+## Available Functions
+
+### String/Character Manipulation
+
+`lowercase` - Lowercases the provided value.<br/>
+**Example**: `{{.FieldName | lowercase}}`
+
+`replace` - Find/replace on the provided value.<br/>
+**Example**: `{{.FieldName | replace - _}}`
+
+`truncate` - truncates the previous value to the provided number of characters.<br/>
+**Example**: `{{.FieldName | truncate 10}}`
+
+`truncate_sha256` - Truncates the previous value to the provided number of characters. The last 8 characters of the
+truncated value will be the first 8 characters of the SHA256 hash of the truncated characters.<br/>
+**Example**: `{{.FieldName | truncate_sha256 20}}`. If `FieldName` is `abcdefghijklmnopqrstuvwxyz`, all characters after
+the 12th (`l`) are removed and SHA256 hashed (`872808ffbf...1886ca6f20`).
+The first 8 characters of the hash (`872808ff`) are then appended to the end of the first 12 characters from the
+original value: `abcdefghijkl872808ff`.
+
+`uppercase` - Uppercases the provided value.<br/>
+**Example**: `{{.FieldName | uppercase}}`
+
+### Generating Values
+
+`random` - Generates a random string from lowercase letters, uppercase letters, and numbers. Must include a
+number indicating how many characters to generate.<br/>
+**Example**: `{{random 20}}` generates 20 random characters
+
+`timestamp` - The current time. Must provide a formatting string based on Go’s time package.<br/>
```

Comment:
If I'm wanting this timestamp in my timezone, I need to provide the formatting string that uses my desired TZ correct? This part of time formatting always throws me. Does this function accept things like `time.RFC3339` ?

Reply:
It does not accept `time.RFC3339` (like: `{{ timestamp time.RFC3339 }}`). I wish it did, but I don't think there is an easy way of handling that. As for timezones, if you specify `Z` I believe it will include your local timezone. In this case, it would be the local timezone of the Vault server. At the moment I don't anticipate this function being used very often. If we get more usage of it with feedback for improvements we can look into it then.

Reply:
👍